### ISSUE TITLE
Java 16 Soft Compatibility: use Java SE libraries & stop module force-opening and too much reflecting

### ISSUE THREAD

**Issue by teddyxlandlee** [McModLauncher/securejarhandler#10]
This is what the force-opening had done:    
```groovy
jvmArgsPrepend = ['--add-opens java.base/java.util.jar=ALL-UNNAMED --add-exports java.base/sun.security.util=ALL-UNNAMED -Dsecurejarhandler.throwOnMissingModule=false']
```
Let's also have a look at `module-info.java`:   
```java
  
import cpw.mods.cl.ModularURLHandler;
import cpw.mods.cl.UnionURLStreamHandler;
import cpw.mods.niofs.union.UnionFileSystemProvider;

module cpw.mods.securejarhandler {
    exports cpw.mods.jarhandling;
    exports cpw.mods.jarhandling.impl;
    exports cpw.mods.cl;
    requires jdk.unsupported;  // This is not a part of Java SE standard libraries. Most of the OpenJDK JREs would abandon this.
    requires org.objectweb.asm;
    requires java.base;
    provides java.nio.file.spi.FileSystemProvider with UnionFileSystemProvider;
    uses cpw.mods.cl.ModularURLHandler.IURLProvider;
    provides ModularURLHandler.IURLProvider with UnionURLStreamHandler;
}
```
Also see `cpw.mods.jarhandling.impl.Jar`:   
```java
public class Jar implements SecureJar {
    private static final CodeSigner[] EMPTY_CODESIGNERS = new CodeSigner[0];
    private static final UnionFileSystemProvider UFSP = (UnionFileSystemProvider) FileSystemProvider.installedProviders().stream().filter(fsp->fsp.getScheme().equals("union")).findFirst().orElseThrow(()->new IllegalStateException("Couldn't find UnionFileSystemProvider"));
    private final Manifest manifest;
    private final /*java.base/sun.security.util.*/ManifestEntryVerifier mev;  // non-exported class. We cannot guarantee its stability and whether it be moved by Oracle or OpenJDKs.
```

Also see `cpw.mods.jarhandler.impl.SecureJarVerifier`:  
```java
    static {
        /* ... */
        try {
            jarVerifier = JarInputStream.class.getDeclaredField("jv");
            sigFileSigners = jarVerifier.getType().getDeclaredField("sigFileSigners");
            existingSigners = jarVerifier.getType().getDeclaredField("verifiedSigners");
            parsingMeta = jarVerifier.getType().getDeclaredField("parsingMeta");
            anyToVerify = jarVerifier.getType().getDeclaredField("anyToVerify");
            jarVerifier.setAccessible(true);
            sigFileSigners.setAccessible(true);
            existingSigners.setAccessible(true);
            parsingMeta.setAccessible(true);
            anyToVerify.setAccessible(true);
        } catch (NoSuchFieldException e) {
            throw new IllegalStateException("Missing essential fields", e);
        }
```  
<p>Using reflect to handle the <b>private</b> methods in <code>java.base/java.util.jar.JarInputStream</code> is not so secure as constructing your own Jar Security API.  </p>  

-------------------------------------------------  

### ~~Something Nonsense~~ Some advice  
<p>Although I'm a Fabric Modder, I am always keeping concerning the development of Minecraft Forge. Before Minecraft 1.17, Forge does not compat Java 9+ (for sun.security.util.* and other package accessibility changes) as well as OpenJ9 VM (?).  
The release of Minecraft Forge 37.0 has started a new era of Minecraft modding cuz Forge <s>has to compat</s> the Java 9 Module System.  
However I am very disappointed at the way handling the solution of the Java 9 issue. Of course JDK includes many things useful, some of them are for JDK inner use and some of them are even deprecated (i.e. `jdk.unsupported/sun.misc.*`). Is it good to expose the "unstable" clothes under the jacket of Java API? If it is good, I can even throw `org.objectweb.asm` away and use `jdk.internal.org.objectweb.asm` and kick `com.ibm.icu4j` off and invoke `jdk.internal.icu` to shrink my third-party libraries.  
See? Using these unstable libraries are harmful to the stability of `securejarhandler` itself. Why don't you <b>construct your own library</b> based on <code>jdk.unsupported</code>, <code>sun.security.util</code>, <code>JarVerifier</code> or something like that. Although Fabric lacks of the functionality, it is stable, from Java 8 to Java 16, from Hotspot to OpenJ9, from Minecraft 1.14 to Minecraft 1.17. In many ways Minecraft Forge is more experienced than Fabric, and I believe Forge can improve its stability and make us mod players more free on playing.  </p>  

------------------------------------------------  

### Some issues on 1.16.x compatibility  
<p>In my opinion, using a Java-9-compatible patch to Minecraft Forge 36 (Minecraft 1.16.x) can benefit 1.16.x players who use Java 16&nbsp;<s> and hate Java 1.8</s>.</p>

**Comment by LexManos** [comment]
> Is it good to expose the "unstable" clothes under the jacket of Java API? If it is good, I can even throw `org.objectweb.asm` away and use `jdk.internal.org.objectweb.asm` and kick `com.ibm.icu4j` off and invoke `jdk.internal.icu` to shrink my third-party libraries. See?

You're just going off on a tangent/slippery slope argument there.
We are not 'exposing' anything. Java already 'exposes' it. We're just using it.
Us using something doesn't 'expose' it to modders any more then it already is.

Forge is stable just fine, you're really walking on thin ice with your advocation/fanboying over Fabric. You do not know what you're actually talking about in regards to the differences in the two projects. Here is your one and only warning. Stop it. 

There are a lot of places that need to use reflection and other internals to do the things that need to be done. Yes it could be cleaned up and probably made to touch a smaller area. But we wernt really in the mood to flat out copy pasta large swafts of JDK code.

If you have actual actionable suggestions we'll entertain them. However, it's one of the many manpower vs priority things that exist in open/community projects such as this.

**Comment by teddyxlandlee** [comment]
Sorry to reply late.  

>     exports cpw.mods.cl;
>     __requires jdk.unsupported;__
>     requires org.objectweb.asm;  

I have double-checked this dependency and found that using `jdk.unsupported` is just fine. The `Unsafe` API have to be used in many ways. Even LWJGL is using it. In this way, Forge is stable across JREs.  

-------------------------------

> jvmArgsPrepend = ['--add-opens java.base/java.util.jar=ALL-UNNAMED --add-exports java.base/sun.security.util=ALL-UNNAMED -Dsecurejarhandler.throwOnMissingModule=false']  

Using this _is_ necessary in developing environment, of course. However, when `securejarhandler` is packed into a jar, maybe adding `Add-Exports` and `Add-Opens` to `META-INF/MANIFEST.MF` is a better solution so that MinecraftForge launcher meta JSONs can shrink these `--add-opens` & `--add-exports` JVM arguments.  
Maybe you'll say this is useless, but if other developers use MinecraftForge/securejarhandler, they don't need to add these annoying args ever.  

---------------------------------

> In my opinion, using a Java-9-compatible patch to Minecraft Forge 36 (Minecraft 1.16.x) can benefit 1.16.x players who use Java 16.

Maybe this is not a huge framework. I think the best solution is using `Add-Exports` and `Add-Opens` in `MANIFEST.MF`, and replace legacy `securejarhandler`s to the latest.  

---------------------------------

Good luck.

**Comment by LexManos** [comment]
Manifest based exports/opens do not work properly with the module system (even tho the spec says it should) in our tests with both executable jars as well as normal module paths.

TLDR the module system is crap and what we do is done because it needs to be done.
If you can provide some actual function tests we can work with I'd be willing to entertain it. However 1/2 the jres out there work one way and the other 1/2 work another. So it would have to be some framework where we can easily test.

**Comment by ZekerZhayard** [comment]
Manifest-based `add-opens` and `add-exports` should be valid only when the `java -jar` command is used, and must be written in the main jar, so it is invalid to write it in securejarhandler.
However, it is most likely to write `Add-Opens` and `Add-Exports` in the 1.16.5 forge server launcher jar, so that you can start the forge 1.16.5 server with only `java -jar forge.jar` in Java 16 without having to add additional parameters. 

This is the launcher jar that I added the corresponding `Add-Opens` and `Add-Exports` to `MANIFEST.MF`. Just extract and replace the installed forge server launcher jar. **Used on your risk**
[forge-1.16.5-36.2.4-repack.jar.zip](https://github.com/MinecraftForge/securejarhandler/files/7149247/forge-1.16.5-36.2.4-repack.jar.zip)

**Comment by LexManos** [comment]
@ZekerZhayard That tells us nothing new. We already know that the manifest functions with executable jars. That has absolutely nothing to do with what we are talking about.
As I said this system DOES NOT work with the module system.
You statement/example has nothing to do with 1.17, SJH, or anything going on right now.

Right now there are two tasks, and two tasks only:
1) Provide a way to get UnSafe without use of reflection/need for additional Opens.
2) Provide a implementation of verification of jar signatures and production of CodeSource values without the violation of OpenJDK's GPL license or the use of reflection.

Those are the main issues at this point.

**Comment by cpw** [comment]
This doesn't seem to be an issue that SJH can fix? It seems you're complaining about the fact that we use AddOpens? This is necessary because the minecraft launcher launches the game using classpath mode - there is no `-jar` we can rely on to get our AddOpens in that way. We do use it for the server as lex noted.

A standalone JarVerifier is a huge amount of development effort, and would be a maintenance nightmare. I looked into doing that and you end up having to implement all of the certificate code as well, because guess what, none of THAT is open either. Adding an opens and exports seems to be far more viable and is the "proper way to do it". Just go look at your favourite IDE running under J11 and tell me that "add opens is bad" mmmkay.
